Modules/_asynciomodule.c

```diff
@@ -598,11 +598,24 @@ future_set_exception(asyncio_state *state, FutureObj *fut, PyObject *exc)
         return NULL;
     }
     if (Py_IS_TYPE(exc_val, (PyTypeObject *)PyExc_StopIteration)) {
-        Py_DECREF(exc_val);
-        PyErr_SetString(PyExc_TypeError,
-                        "StopIteration interacts badly with generators "
-                        "and cannot be raised into a Future");
-        return NULL;
+        const char *msg = "StopIteration interacts badly with "
+                          "generators and cannot be raised into a "
+                          "Future";
+        PyObject *message = PyUnicode_FromString(msg);
+        PyObject *err = PyObject_CallOneArg(PyExc_RuntimeError, message);
+        if (err == NULL) {
+            Py_DECREF(message);
```

Comment:
Move the DECREF up one line, before the `if` -- the message always needs to be freed, and the call adds a reference to it.